mypy/semanal.py

```diff
@@ -4542,9 +4542,15 @@ def current_symbol_table(self, escape_comprehensions: bool = False) -> SymbolTab
         if self.is_func_scope():
             assert self.locals[-1] is not None
             if escape_comprehensions:
+                assert len(self.locals) == len(self.is_comprehension_stack)
+                # Retrieve the symbol table from the enclosing non-comprehension scope.
                 for i, is_comprehension in enumerate(reversed(self.is_comprehension_stack)):
                     if not is_comprehension:
-                        names = self.locals[-1 - i]
+                        if i == len(self.locals) - 1:  # The last iteration.
+                            # The caller of the comprehension is in the global space.
+                            names = self.globals
+                        else:
+                            names = cast(SymbolTable, self.locals[-1 - i])
```

Comment:
Why is this `cast()` needed? Could you use a `# type:` comment on the first assignment instead?

Reply:
the cast here was because `self.locals` is a list of `Optional[SymbolTable]`, and so without it, you would get:
```
mypy\semanal.py:4552: error: Incompatible types in assignment (expression has type "Optional[SymbolTable]", variable has type "SymbolTable")  [assignment]
                                names = self.locals[-1 - i]
                                        ^
```

I reasoned (at the time) that it is a safe cast as long as we are not retrieving the `None` that is the first element in the list.
https://github.com/python/mypy/blob/ed9ba4f87fadefa9de335246bf2c12135dec735c/mypy/semanal.py#L226

However, now that you have mentioned to check the class scope, I have realised that there may be other elements in the list that are None: https://github.com/python/mypy/blob/ed9ba4f87fadefa9de335246bf2c12135dec735c/mypy/semanal.py#L1166
Perhaps once I deal with the class case properly this cast will not be necessary. Either way, for now, my original assumption that _only_ the first entry is None seems invalid! So I will find an alternative approach.

Reply:
Since comprehensions in a class scope e.g. 
```
class Example:
    [(j := i) for i in range(5)]
```
can also result in the retrieved `self.locals[-1 -i]` symbol table being None, I added extra logic for that case, which removed the need for the cast, in https://github.com/python/mypy/pull/9062/commits/a68bd68958cfa37bfc1a4d9d9946c07b3a63e518.

Because this specific case is called out in https://www.python.org/dev/peps/pep-0572/#scope-of-the-target as being invalid syntax (I know that you know this, I am just being verbose), the specific logic that's added is actually just an assertion that we don't end up in that case.